### ISSUE TITLE
Marshal.ReleaseComObject does not account for ComWrappers

### ISSUE THREAD

**Issue by kant2002** [dotnet/runtime#54317]
I try to make WPF work using ComWrappers only, and hit followin road block. Currently RCW in WPF released using `Marshal.ReleaseComObject`. This method does [not take into account ](https://github.com/dotnet/runtime/blob/89313ca06932bdb69c53965ec1e37b67f920ee95/src/coreclr/System.Private.CoreLib/src/System/Runtime/InteropServices/Marshal.CoreCLR.cs#L487) that some code is using ComWrappers

I think in spirit of #50500 that behaviour should change.
/cc @AaronRobinsonMSFT

**Comment by AaronRobinsonMSFT** [comment]
@kant2002 This is by design. The appropriate API to call instead is [`Marshal.Release`](https://docs.microsoft.com/dotnet/api/system.runtime.interopservices.marshal.release) or some other variant to do the decrement. The `Marshal.ReleaseComObject` API is targeted for manipulating the built-in COM interop wrappers not user defined wrappers (that is, `ComWrappers`). What is the intent of calling this API on a `ComWrappers` instance?

**Comment by kant2002** [comment]
I'm hit that when experimenting how far can I go with ComWrappers in WPF (to see NativeAOT ability potentially)

WPF retreive object using `DllImport`, and it is created using `ComWrappers`, then it is assumed that this is `__ComObject` and using `Marshal.ReleaseComObject` to release instance. I assume this is would be common pattern for existing COM codebases. 

This is quick summary on the libraries which use this functions. I count only files, because it's easier to do in the browser.

https://github.com/dotnet/winforms/search?q=ReleaseComObject 6 non test files
https://github.com/dotnet/wpf/search?q=ReleaseComObject 30 no test files

https://github.com/dotnet/runtime/search?q=ReleaseComObject
System.DirectoryServices - 1 file
System.Data.OleDb - 7 file
System.Speech - 9 file
System.Drawing.Common - 1 file (Icon class)
System.DirectoryServices.AccountManagement - 1 file
System.Management - 4 file

Other files.
libraries/Common/src/System/Runtime/InteropServices/ComEventsSink.cs - 1
tools/Common/TypeSystem/Ecma/SymbolReader/UnmanagedPdbSymbolReader.cs - 1

WinForms almost not impacted if not using ActiveX. It works perfectly fine for me in NativeAOT in relatively complex app.
WPF not even starts with ComWrappers attached.
DirectoryServices, Management, Speech based on previous discussion are less supported. I would cry only about Management because admin work, and Speech because it's fun. I always can contribute if would be desperate.

**Comment by AaronRobinsonMSFT** [comment]
@kant2002 Rather than updating this API I would recommend converting entirely to `ComWrappers` and the returned wrapper could implement `IDisposable` pattern. From within that I would release as desired. I don't think updating that API to handle `ComWrappers` wrappers is appropriate because there isn't an obvious way to convert a user defined wrapper into something we could release &ndash; it is an implementation detail of a `ComWrappers` instance.

**Comment by jkotas** [comment]
I agree that the ultimate goal should be to get to 100% ComWrappers, but I think it would make sense to update APIs like `Marshal.ReleaseComObject` to handle ComWrappers to make migration to ComWrappers easier.

Supporting ComWrappers in APIs like `Marshal.ReleaseComObject` costs very little and it is in the same spirit as `ComWrappers.RegisterForMarshalling` APIs that was also introduce to ease the migration.

**Comment by AaronRobinsonMSFT** [comment]
> Supporting ComWrappers in APIs like Marshal.ReleaseComObject costs very little and it is in the same spirit as ComWrappers.RegisterForMarshalling API that was also introduce to ease the migration.

I don't think it costs very little. It is a complicated scenario with a fair bit of nuance.

1) This could only be made to work for a globally registered `ComWrappers` instance for the marshalling scenario. Which means we would need to detect that one was registered _and_ the object passed in was created with it.
1) We would then call the `ReleaseObjects()` on the `ComWrappers` instance and pass in a single object &ndash; which is far more expensive than the current API implementation.

The above would mean (1) when and how this works is complicated and (2) the performance relative to the existing behavior (i.e., built-in) is much more and that defeats the purpose of `ComWrappers`.

I would argue we should separate them at this point and rely on either an `IDisposable` contract or perhaps something else. I do not see the value in overloading this API when it has a fair bit of complexity that users are bound to get wrong and shouldn't really be using anyways. The number of times we've had to diagnose early release RCWs is too many to count for me and introducing that outside of the `ComWrappers` API seems going down a road that has already be tread with many lessons learned.

**Comment by AaronRobinsonMSFT** [comment]
@kant2002 I took another look at this and really am concerned about the complexity here. We would need to update both `Marshal.ReleaseComObject` and `Marshal.FinalReleaseComObject` &ndash; the behavior would basically be identical. Aside from the relative cost of implementing these APIs we also must rely on a user's `ComWrappers` implementation. The contract for these APIs is rather specific and users have come to rely on an actual release &ndash; we can guarantee that with the built-in interop. In this new world we would rely on the `ComWrappers` implementation to provide a `ReleaseObjects()` that disables the wrappers safely and doesn't just return.

An argument could be made this is an advanced scenario so that is the cost. I'm sympathetic to this perspective actually. However, the arguments I am hearing are around successfully transitioning from the built-in system. This means we want to guarantee semantics to some degree and I don't see how we can do that. I will say that as we move forward with `ComWrappers` a convention for how wrappers are written should be defined. I personally like `IDisposable` but there are a lot of options here.

I would recommend doing the following prior to these calls:

```csharp
if (obj is __ComObject)
    Marshal.ReleaseComObject(obj);
```

**Comment by jkoritzinsky** [comment]
@AaronRobinsonMSFT `System.__ComObject` is internal to CoreLib, so that’s not possible.

**Comment by AaronRobinsonMSFT** [comment]
> @AaronRobinsonMSFT System.__ComObject is internal to CoreLib, so that’s not possible.

Doh! All those years of seeing that in the VS debugger and I never realized it was marked `internal`. Hmmm. Now I think we need to talk about this...

**Comment by kant2002** [comment]
I'm fine with either IDisposable, finalizer or other custom interface, but I would like that somehow this function work almost like it is expected in that function. 
Maybe behind some switch. Without that, you have to go for ComWrappers and rewrite all P/Invoke calls, and wait for libraries to catch up.
Essentially this prevents global ComWrappers to be useful for existing applications in my opinion.

I personally think about this interface
```
interface IComReference
{
   int Release();
}
```

**Comment by jkotas** [comment]
> Essentially this prevents global ComWrappers to be useful for existing applications in my opinion.

The global ComWrappers instance is designed to only able to cover the simplest COM interop cases transparently (simple interface marshaling). It won't cover existing libraries and application that use COM interop in more advanced ways. It is very likely that applications or libraries that are using COM interop heavily will require changes to work well with ComWrappers (either global or local instance).

> if (Marshal.IsComObject(obj))
>    Marshal.ReleaseComObject(obj);

This will prevent the app from crashing with ComWrappers, but it won't make it work well. The code was presumably explicitly releasing the object for a reason, and not just because of it can. I think it would need to be:

```
if (Marshal.IsComObject(obj))
    Marshal.ReleaseComObject(obj);
else
    (obj as IDisposable)?.Dispose();
```

and the respective ComWrappers would have to implement `IDisposable`.

**Comment by kant2002** [comment]
> I think it would need to be:
That's seems pragmatic. I see how it can be easy applied across codebases.

If nothing would be added/changed, I can provide separate PR's for each library in this repo. Any objections?

**Comment by jkotas** [comment]
The challenge with changes like this is testing - how do we know that it actually works?

I think it would prefferable to work towards the 100% ComWrappers solution: Pick one of these libraries, figure out how to make wean it off of the built-in COM support by default, and make it happen. Rinse & repeat.

**Comment by kant2002** [comment]
Let's see. I would pick System.Drawing.Common as easiest target with only 2 interfaces. IStream and IPicture.
How I create RCW for IPicture, and CCW for IStream? Should I create them manually, or should I use codegen of some form? This one seems easy to create manually, but what about other libraries?

**Comment by AaronRobinsonMSFT** [comment]
@kant2002 I believe that @eerhardt is also looking into System.Drawing.Common. You two might want to sync.

**Comment by eerhardt** [comment]
Yes, I started working on moving System.Drawing.Common over to COM wrappers in order to make it trim compatible. Here are the current ILLink warnings in Drawing:

https://github.com/dotnet/runtime/blob/cff5854fb02259b88cab89f77e8d403809300ae8/src/libraries/System.Drawing.Common/src/ILLink/ILLink.Suppressions.xml#L16-L81

They are all:

> IL2050: Trim analysis: Correctness of COM interop cannot be guaranteed
P/invoke method 'method' declares a parameter with COM marshalling. Correctness of COM interop cannot be guaranteed after trimming. Interfaces and interface members might be removed.

Moving them to use COM wrappers will resolve these ILLink warnings.

**Comment by eerhardt** [comment]
@kant2002 - if you want, you could pick up one of the other libraries?

https://github.com/dotnet/runtime/search?q=IL2050

* System.Management
* System.DirectoryServices
* System.Data.Odbc
* System.Data.OleDb

**Comment by kant2002** [comment]
@eerhardt For now I would like to shadow you, to see how ComWrappers should be implemented. Can you tag me if you have PR or draft? Is there general consensus how it should be done? I can start poking into System.Data.Odbc or System.Management based on what should be done. Want to limit scope to easiest target first.

**Comment by eerhardt** [comment]
For sure! I'll tag you when I have a PR ready. Probably sometime next week.

**Comment by eerhardt** [comment]
FYI - the first round of converting System.Drawing.Common to ComWrappers: https://github.com/dotnet/runtime/pull/54636.

**Comment by kant2002** [comment]
Does System.Speech not appear in search for IL2050 because it uses `ComImport` attribute for object creation and no built-in COM marshalling is used, just built-in object creation?

**Comment by eerhardt** [comment]
> Does System.Speech not appear in search for IL2050 because it uses ComImport attribute for object creation and no built-in COM marshalling is used, just built-in object creation?

No, we choose not to analyze System.Speech for trimmability yet, because it isn't a very used assembly (it is only getting supported on .NET Core in the 5.0 time frame). Here is the list of assemblies we ignore for trimmability (for now):

https://github.com/dotnet/runtime/blob/bde0322a9e11fba40a7a65c6d38200d6514efc72/src/libraries/illink-oob.targets#L15-L25

Some of that reasoning is listed here: https://github.com/dotnet/runtime/pull/52272#discussion_r627140855

**Comment by AaronRobinsonMSFT** [comment]
I think the follow up here should be documentation regarding contracts and how implementers should consume `ComWrappers`.
